File: data_structures/javascript/DoublyLinkedList.js
class DoublyLinkedNode {
    constructor(data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLinedList {
    constructor() {
        this.head = null;
        this.tail = null;
        this.length = 0;
    }

    //O(1) time
    add(data) {
        const newNode = new DoublyLinkedNode(data);

        if (!this.length) {
            this.head = newNode;
            this.tail = newNode;
        } else {
            this.tail.next = newNode;
            newNode.prev = this.tail;
            this.tail = newNode;
        }
        this.length++;
        return newNode;
    }

    //O(this.length) time
    findNodeAtPosition(position) {

        if (position > this.length || position < 1 || this.length === 0) {
            throw new RangeError(`Position ${position} is not available in LinkedList of length ${this.length}`);
        }

        let currentNode = this.head;
        let currentPosition = 1;

        while (currentPosition < position) {
            currentNode = currentNode.next;
            currentPosition++;
        }
        return currentNode;
    }

    //O(this.length / 2) time
    removeNodeAtPosition(position) {

        if (position > this.length || position < 1 || this.length === 0) {
            throw new RangeError(`Position ${position} is not available in LinkedList of length ${this.length}`);
        }

        const medianPosition = Math.floor(this.length / 2);
        let currentNode, previousNode, nextNode, nodeToDelete, deletedNode;

        if (position >= medianPosition) {
            currentNode = this.tail;
            currentPosition = this.length;

            while (currentPosition > position) {
                currentNode = currentNode.prev;
                currentPosition--;
            }

        } else {
            currentNode = this.head;
            currentPosition = 1;

            while (currentPosition < position) {
                currentNode = currentNode.next;
                currentPosition++;
            }
        }

        previousNode = currentNode.previous;
        nodeToDelete = currentNode;
        nextNode = currentNode.next;

        previousNode.next = nextNode;
        nextNode.previous = previousNode;
        deletedNode = nodeToDelete;
        nodeToDelete = null;
        return deletedNode;

    }

    //TODO: test remove at position; write findNodeWithData, removeNodeWithData;
}

File: data_structures/ruby/doubly_linked_list.rb
class DoublyLinkedListNode
    attr_accessor :data, :prev, :next
    def initialize(data)
        @data = data
        @prev = nil
        @next = nil
    end
end

class DoublyLinkedList
    def initialize
        @head = nil
        @tail = nil
        @length = 0
    end

    def add(data)
        new_node = DoublyLinkedListNode.new(data)

        if @length == 0
            @head = new_node
            @tail = new_node
        else
            @tail.next = new_node
            new_node.prev = @tail
            @tail = new_node
        end
    end

    def find_node_at_position(position)
        if (position > @length || position < 1)
            raise ArgumentError, "Position must be within range(0, #{@length}) or list is currently empty"
        end

        current_node = @head
        current_position = 1

        until current_position == position
            current_node = current_node.next
            current_position += 1
        end

        return current_node
    end

    def find_node_with_data(data)
        if @length == 0
            return nil
        end

        current_node = @head
        until current_node.next.nil? || current_node.data == data
            current_node = current_node.next
        end
        if current_node.data == data
            return current_node
        else
            return nil
        end
    end

    def remove_node_at_position(position)

        if (position > @length || position < 1)
            raise ArgumentError, "Position must be within range(0, #{@length}) or list is currently empty"
        end

        current_node = nil
        previous_node = nil
        next_node = nil
        node_to_delete = nil
        deleted_node = nil

        if (position == 1)
            node_to_delete = @head
            @head = node_to_delete.next
            deleted_node = node_to_delete
            return deleted_node
        elsif (position == @length)
            node_to_delete = @tail
            @tail = node_to_delete.prev
            deleted_node = node_to_delete
            node_to_delete = nil
            return deleted_node
        else
            until current_position == position
                current_node.next = current_node
                current_position += 1
            end
            previous_node = current_node.prev
            node_to_delete = current_node
            next_node = current_node.next

            previous_node.next = next_node
            next_node.prev = previous_node
            deleted_node = node_to_delete
            node_to_delete = nil
            @length -= 1
            return deleted_node
        end
    end

    def delete_node_with_data(data)
        if @length == 0
            return nil
        else 
            current_node = @head

            until current_node.next.nil? || current_node.data == data
                current_node = current_node.next
            end
            
            if (current_node.data == data)
                previous_node = current_node.prev
                node_to_delete = current_node
                next_node = current_node.next

                previous_node.next = next_node
                next_node.prev = previous_node
                deleted_node = node_to_delete
                node_to_delete = nil
                @length -= 1
                return deleted_node
            else
                return nil
            end
        end
    end
end